### DISCUSSION TITLE
How does uxarray compares with other ugrid libraries?

### DISCUSSION THREAD

**Discussion by ocefpaf** [UXARRAY/uxarray#421] [General]
This is not really a feature request, maybe it is better as a topic discussion? Anyway here is the gist of it:

I'm taking a tour at the many ugrid libraries out there

- ugridpy: https://github.com/Deltares/UGridPy
- xugrid: https://github.com/Deltares/xugrid
- gridded: https://github.com/NOAA-ORR-ERD/gridded
- pyugrid: https://github.com/pyugrid/pyugrid

Ugridpy depends on two c++ libraries that are quite hard to compile and upstream only provides Windows binaries.

Xugrid seems to be the most "feature complete" but it failed to load a compliant dataset (I'm still investigating that).

Gridded has almost all the functionalities of xugrid but operates on netcdf4-python objects instead of xarray.

Pyugrid is the oldest and also only operates on netcdf4-python.

Uxarray can read from multiple grid formats, only from files for now (PR #68 fixes that), but lacks some of the xugrid functionalities like plotting and high level slices.

I'd love to hear thoughts from the devs and/or comments/corrections on my assent above.

**Comment by paullric** [comment]
I'd say that UXARRAY is most similar to xugrid.  The expectation is that all code will be in Python and built upon the xarray library.  A key focus of this effort is on high performance algorithms for very large unstructured grids (i.e., more than 1M faces).  Ultimately we want users to be able to interact with unstructured data without necessarily being aware that it is unstructured -- that is, providing all of the functionality one would expect from structured data operators.  As such we're trying to be fairly ambitious with our suite of operators (pointwise/areal sampling, zonal/meridional means, cross-sections, regridding, visualization, shapefile support, etc.).

**Comment by dcherian** [comment]
Iris has experimental ugrid support too now: https://scitools-iris.readthedocs.io/en/stable/generated/api/iris/experimental/ugrid.html#module-iris.experimental.ugrid

**Comment by veenstrajelmer** [comment]
@ocefpaf I wonder about the dataset you were trying to load with xugrid. xugrid is cf-compliant so it should work if the dataset is also cf-compliant. If it is not, it should also still be fairly simple to use the dataset. Could you test it again or share it? Or is a comparable file available in your [testdata](https://github.com/UXARRAY/uxarray/tree/main/test/meshfiles) folder?

Two additional reasons why I use xugrid over uxarray:
- code quality and unit test coverage (I know I can trust xugrid developments will not break my code)
- support for partitioned grids, including merging them
- lean dependencies, but powerful and efficient either way
- designed closely next to xarray, so most of the things are super intuitive (albeit calling a `ugrid` accessor on the way)
- numerous features (like reindexing, rasterizing, cross sections, regridding, etc) I really appreciate. I have not dived into uxarray yet so those might be there also, but the docs seem to mainly cover visualization of data.

I think teaming up could be useful for both packages, at least sharing the lessons learned about for instance performance.

**Reply by philipc2** [reply]
Hi @veenstrajelmer 

Thanks for chiming in! Would be great to sit down and catch each other up on the development of both packages, since there is definitely a common interest. We've recently revamped our [User Guide](https://uxarray.readthedocs.io/en/latest/userguide.html) if you'd like to learn more about our functionality.

**Reply by ocefpaf** [reply]
@veenstrajelmer this was 2 years ago. I'll try to find what it was and why I did not follow up on the tests. However, I can assure you I only tested datasets that were 100% CF and UGRID compliant.
